lib/mediawiki/butt.rb

```diff
@@ -75,6 +82,18 @@ def post(params)
 
       response = JSON.parse(@session.post(@url, params).body)
 
+      # If our tokens have expired, clear them, re-set the relevant token argument, and re-call this #post method
+      if response.dig('error', 'code') == 'badtoken' && @first_token_try
+        @tokens.clear
+        token_param = params.keys.select { |i| i.to_s.end_with?('token') }[0]
+        params[token_param] = get_token(@prev_token_type)
+        # Prevent token retry attempt in next #post call
+        @first_token_try = false
+        return post(params)
+      end
+      # Reset token retry value if no badtoken error has occurred.
+      @first_token_try = true
```

Comment:
The `@first_token_try` variable is something I wasn't totally sold on, but I couldn't think of a better way to handle it. If you have any ideas @xbony2 let me know and I can rewrite that part.